### DISCUSSION TITLE
Stack-like ctrl-tab and ctrl-shift-tab behaviour

### DISCUSSION THREAD

**Discussion by byildiz** [ghostty-org/ghostty#2393] [Feature Requests, Ideas]
Stack-like `ctrl-tab` and `ctrl-shift-tab` behaviour as `cmd-tab` (`alt-tab` for linux and windows) behaves can be more useful. We can cycle through lastly used tabs easily. Firefox and iTerm use same approach. As an example, let's have a ghostty window with 4 tabs opened in a row, we have a list of tabs:
```
[1, 2, 3, 4]
```
When user hits one `ctrl-tab`, _tab 3_ will be opened and the list will be:
```
[1, 2, 4, 3]
```
At the last state after two `ctrl-tab` hits (without releasing `ctrl`), _tab 2_ will be opened and the list will be:
```
[1, 4, 3, 2]
```
One `ctrl-shift-tab` opens _tab 1_ and the list will be as follows:
```
[4, 3, 2, 1]
```
In summary, `ctrl-tab` cycles through the lastly used tabs and `ctrl-shift-tab` cycles through it reverse. 

I recorded iTerm window to simulate above example:

https://github.com/user-attachments/assets/62384319-14bb-4a26-ac92-ecbb56f00124

**Comment by mitchellh** [comment]
I converted this to a discussion since I'm not fully ready to accept this feature idea yet. We've bound `ctrl+[shift]+tab` by default to cycle right/left since that seems to be the OS default or idiomatic binding. I'm not opposed to introducing a new stack-like action but want to consider the design more before converting back to an issue!

**Reply by lyallcooper** [reply]
<kbd>Ctrl</kbd>+<kbd>Tab</kbd> going to the next tab to the right is definitely the more common default behavior on macOS, but having the option to change it to last opened tab instead (à la Firefox and iTerm as mentioned) would be hugely to me at least.

**Reply by BenSandeen** [reply]
This is one of the two things stopping me from immediately migrating.  The other is not being able to use Cmd+F/Ctrl+F for finding text

**Reply by ArbiterOfBuffoonery** [reply]
I advocate for having the `next_tab` and `previous_tab` actions function as they do today, and adding new actions (e.g. `next_tab_mru` and `previous_tab_mru`) that allow keybindings to trigger this functionality.

That way those who prefer the stack switching can change the default binding, but the idiomatic defaults will be the zero config state.

**Reply by BenSandeen** [reply]
If I understand your suggestion, that would enable me to just tweak a config and then the normal "next tab" and "previous tab" keybindings will work in an MRU fashion?  I don't care how it's implemented behind the scenes, as long as I don't have to change the keypresses I use.  Maybe an ideal solution (with a lot of flexibility) would be to enable users to configure `next_tab`, `previous_tab`, `next_tab_mru`, and `previous_tab_mru`, _and_ add a setting called something like `tab_cycle_order` which can accept two options, something like `in_order` and `most_recently_used`.  That way people could configure both pairs of keybindings if they wanted, but people like me could just switch the `tab_cycle_order` to the MRU option and be content

**Reply by ArbiterOfBuffoonery** [reply]
I also want mru behavior, so my plan was when the `next_tab_mru` and `previous_tab_mru` options were available (in theory), I was going to bind them to `ctrl+[shift]+tab` and then leave the `previous_tab` and `next_tab` unbound (or maybe bind them to something else like `ctrl+[right]+left`

I think that means there is no need for a `tab_cycle_order` option in this scenario, but I could be missing something.

**Reply by BenSandeen** [reply]
There may not be a _need_ for it, but it would be a nice convenience function, making configuration for newcomers a bit simpler.  Also, it makes sense to pair them, because most people would want to configure either both or neither of them

**Reply by frammaa** [reply]
I think having to rebind shortcuts makes for a very poor experience. If I look at how it's implemented in Firefox, which is a simple checkbox, this should just be a matter of setting an option to true, similar to the `tab_cycle_order` option that @BenSandeen mentioned. I was imagining something like `cycle-tabs-by-recent = true`, but I guess the `tab_cycle_order` option leaves space for future alternatives.

I agree that this is quite a big deal to me ctrl-tab is integral to my flow so I'd love to see this turned into an issue, in which case I'd be happy to take a stab at implementing it myself, sounds like an approachable first issue.

**Reply by 00-kat** [reply]
> future alternatives

What other orders are common?  I'm only really aware of left-to-right and most-recently-used orders.

**Reply by frammaa** [reply]
I'm also only aware of those two, but I could imagine that users whose language is right-to-left might have different expectations. Would be nice to allow for future flexibility, and if it's never necessary it's not so bad to just have the two options

**Reply by cezdro** [reply]
> I also want mru behavior, so my plan was when the next_tab_mru and previous_tab_mru options were available (in theory), I was going to bind them to ctrl+[shift]+tab and then leave the previous_tab and next_tab unbound (or maybe bind them to something else like ctrl+[right]+left
> 
> I think that means there is no need for a tab_cycle_order option in this scenario, but I could be missing something.

Stack-like Ctrl+Tab has a different behavior if the Control key is kept being released between presses. It is hard to achieve through keybindings alone, although GNOME somehow does it — you can bind Alt+Tab to "Switch windows" instead of "Switch applications" and it will still work like that.

> We've bound ctrl+[shift]+tab by default to cycle right/left since that seems to be the OS default or idiomatic binding.

It used to be the default behavior in all applications before Google made Ctrl+Tab behave in the same way as Ctrl+PgDn in Chrome for some inexplicable reason. Then every browser copied this behavior and every new terminal application works that way too. It definitely is not a default nor idiomatic behavior in OS'es and other applications though. In GNOME, Xfce, Windows and I think KDE too, Alt+Tab switches between applications in the most recent used order by default. I don't use macOS so I can't speak about it.

**Reply by julien-may** [reply]
I'd also love to add my voice in support of MRU tab switching. I feel that for many developers, especially those used to productivity-focused applications (like editors, IDEs, or whatever tab supported applications), this is an integral navigation pattern. Quickly toggling between the last two tabs with Ctrl+Tab without thinking about position is deeply ingrained in daily workflows.

One way this could be introduced without disturbing current users would be to offer a setting like:

```
tab_switching_order = "ltr" # left-to-right or mru 
```

This would allow users to opt into the behavior that matches their workflow, while preserving the current default.

**Reply by dmik** [reply]
For possible reasoning on why this feature should be added, please see my separate comment below.

Regarding the idiomatic way of switching. I cannot speak for all OSes, but on macOS there are two traditional pairs of bindings that mirror each other in many native apps: `Shift+Ctrl+Tab` / `Ctrl+Tab` and `Shift+Cmd+[` / `Shift+Cmd+]`. Both pairs switch tabs left (prev) / right (next), accordingly. These bindings work even in Finder (Safari, Terminal, Preview). Ghostty behaves the same way on macOS, btw. So there is one easy solution to this problem (at least on macOS): keep `Shift+Cmd+[` / `Shift+Cmd+]` bound to `prev_tab` / `next_tab` (which is also consistent with the physical key locations) and add a pair of new actions `lru_tab` / `mru_tab` bound to `Shift+Ctrl+Tab` / `Ctrl+Tab`, as well as a new option `mru-tab-order` set to `false` by default. Then those who want true MRU order for `lru_tab` and `mru_tab` actions (acting as `prev_tab` / `next_tab` synonyms when this option is `false`), will simply set it to `true` and be happy 😊 (and productive).

**Reply by mnemnion** [reply]
> I'm also only aware of those two, but I could imagine that users whose language is right-to-left might have different expectations. Would be nice to allow for future flexibility, and if it's never necessary it's not so bad to just have the two options

At least for macOS, this shouldn't be necessary to configure, provided that the Swift chrome is written appropriately.  This isn't something I'm at all expert in, but I'll drop a [link to this WWDC video](https://developer.apple.com/videos/play/wwdc2022/10107/), the TL;DR being that properly-engineered Mac applications will mirror this kind of thing based on the language, or maybe an additional global setting. The key is that "left" and "right" are absolute, while "forward" is only left for a left-to-right language.

**Comment by 00-kat** [comment]
#1844 is very related and a possible duplicate.

**Comment by dmik** [comment]
https://github.com/ghostty-org/ghostty/discussions/4904 is also a duplicate of this.

Regarding the feature. Being able to switch tabs in MRU order is crucial if there are more than 2-3 tabs and I need to switch between them regularly. Switching in MRU order is much more natural than any other way because it allows to naturally switch contexts of my work like this: "bring me to the thing I was working on before, then bring me back" (or alternatively "...two steps ago and back" etc).

By "much more natural" I mean that the (single-threaded) human brain switches its contexts in MRU order as well, there is just no native concept of numbering brain tasks/contexts or maintaining their "left-right" order, nothing like that at all. In order to go left-right (or Cmd/Ctrl+numbers) the brain has to involve the visual subsystem (eyes) to map between its internal stack-like contexts and their visual linear representation (the left-right order of tabs in the UI). And doing these mappings not only takes a lot of precious brain time (the visual subsystem is a complex thing), it also distracts its resources from working on the tasks themselves because, well, it's a single-threaded organ. (For me personally it is also literally painful to go through such distractions among other things).

Given how easily this feature can be implemented, I really believe it is a must for a product targeting professionals in the first place. I would appreciate it a lot to see this feature implemented in an upcoming update.

Thank you for your great work, it's a decent terminal app (almost my default one — and will become such as soon as MRU order is there).

PS: For a possible implementation, please see my comment in the message thread above.

**Comment by m4dc4p** [comment]
Firefox supports this idea (they call it "[cycle tabs in recently used order](https://support.mozilla.org/en-US/kb/tab-preferences-and-settings#:~:text=Ctrl%2BTab%20cycles%20through%20tabs%20in%20recently%20used%20order%3A)"); Chrome famously DOES NOT (https://issues.chromium.org/issues/40445018 - TL;DR closed as "WontFix").

The benefit of the feature is being able to toggle between two tabs with the same keyboard shortcut. Example:

* In tab 1-3, I have unrelated logs and junk open
* _in tab 4_, I have an editor open
* In tabs 5-8, I have more unrelated stuff
* _in tab 9_, I have a hot-reloading compiler that's watching files being worked on in tab 4.

With "stack like behavior," or "Most Recently Used" (MRU) ordering, where cmd-tab switches to the most recently viewed tab first, I can toggle between tab 4 and tab 9 really fast.

Without MRU, switching between tab 4 and 9 requires multiple cmd-tab keypresses, or that I drag tabs around. Even if the tabs next to each other, I still have to use cmd-tab and cmd-shift-tab to toggle between them.

**Comment by prajjwal000** [comment]
I might be missing something but isn't the behavior of pressing ctrl+tab multiple times for cycling to specific mru tab not supported in trigger sequences.

For example:
If we make an action like mru_tab:position_number . We bind ctrl+tab to mru_tab:1 and ctrl+tab>ctrl+tab to mru_tab:2.
Then pressing ctrl+tab just once would just wait indefinitely for next ctrl+tab and nothing would happen.

https://ghostty.org/docs/config/keybind/sequence

**Reply by prajjwal000** [reply]
If the fancy behaviour isn't even possible, then I think we should implement a mru_tab action and move on

**Comment by FlorianLoch** [comment]
Any change of mind regarding this?

To me that is a very basic, though very effective pattern. Getting back to last active tab requiring a "full cycle" feels very inefficient to me. Especially, when many tabs are open. 

At least for me the most common reason to switch tabs is to go back and forth. I can work around this by accessing the tabs via their index - but that isn't really handy as back-and-forth becomes dependent of the current position.

JetBrains' IDEs, VS Code, Firefox, Vivaldi, Opera, Arc, iTerm etc. supporting this pattern backs my impression that it is quite common in the power-user context.

I am not saying it should be the default setting - but, IMHO, it should be an option for those preferring it.

**Reply by joelvdvoort** [reply]
+1. I regularly recommend ghostty to my peers but don't use it myself because mru_tab feature is sorely missed. I'm still on iTerm because of this.

**Comment by jkinred** [comment]
I sorely missed rapid switching between last active tabs (e.g. 1 -> 7 -> 1 -> 7) using tmux-like `<Prefix><Prefix>` which is `ctrl+space>ctrl+space` in my config. This is by far my most used keybind.

Although some duplicates mention it, they are all pointing back at this discussion where I didn't see it explicitly mentioned.

This discussion seems to be aiming for something more sophisticated, but the basic use case is much simpler. I had an agent implement `last_active_tab` (on [this fork](https://github.com/jkinred/ghostty)) which I'm running locally on Linux and I'm back to tab switching bliss.

I don't intend to submit a PR or start a new discussion because there's already been a lot of duplication and the focal point is this more sophisticated behaviour, but if there were interest in a simpler `last_active_tab` feature I'd learn some Zig, review and tidy up the agents work and submit a PR.

**Reply by jkinred** [reply]
I'm maintaining my fork against main and also maintaining a branch against the latest tagged release (v1.3.1 at time of writing) -- which is what I'm running myself.

https://github.com/jkinred/ghostty/tree/v1.3.1-with-last-active-tab

**Comment by WalterClementsJr** [comment]
in Vim, `:buffer #` to switch to the last active buffer got to be one of my most used keybinds. I think I can speak for many users after seeing duplicate discussions that most of us don't really need a full MRU.

**Reply by jkinred** [reply]
Not sure why this is getting downvoted. Could those downvoting help the discussion by stating why?

**Reply by 00-kat** [reply]
They have nothing backing up their claim that “most of us don't really need a full MRU”, yet want to “speak for many users”. Also: as someone who got a notification for every single discussion for a year, and after I gave up with that still frequently checked our Discord webhook to catch new discussions, I have no clue what duplicate discussions they refer to: I remember exactly one (maybe it was two), and they wanted full MRU.

Based on upvotes/comments here and #1844, the majority seem to want MRU and not switch-to-last-tab. But even then, I really feel like this discussion should be about MRU only, and #1844 should be about switch-to-last-tab only, since:
  - That's far easier to do UI-wise, it's just a keybind.
  - That doesn't need more discussion, unlike this. Issues are for things that don't need feature design.
  - The majority here want MRU; it's somewhat mixed there but the majority there want to stay with what the OP of that wanted.

**Edit**: I decided I might as well action on it: see [my next comment here](https://github.com/ghostty-org/ghostty/discussions/2393#discussioncomment-16754746), and my comment at https://github.com/ghostty-org/ghostty/issues/1844#issuecomment-4343040442.

**Reply by 00-kat** [reply]
> as someone who got a notification for every single discussion for a year […] I have no clue what duplicate discussions they refer to

I see I was very wrong, lol. There was exactly one that wanted full MRU: #4904. The [others](https://github.com/ghostty-org/ghostty/discussions?discussions_q=2393):
  - #11617
  - #9300
  - #10320 (is ambiguous, they demonstrate both apps that use MRU and apps that use switching to the last tab)
  - #10317
  - #10021
  - #4260
  - #5391

Half of them were from a week or so before I subscribed to all notifications, and the rest were in the three month period after I gave up and unsubscribed, so it makes sense that I never saw them, lol.

Assuming every single person who upvoted a comment (here or in #1844) upvoted every other one too, it seems to be around 50/50 (just a guess, I didn't count exactly). In reality it's not the exact same people every time (but I have no proof), so MRU is slightly more popular; but MRU almost entirely subsumes switching to the last tab, so there's bias even there since advocates of switching to the last tab might have expressed support for MRU too. I'm hence inclined to call it 50/50.

**Reply by jkinred** [reply]
Thanks for clarifying the downvoting and doing all that digging @00-kat.

I agree with your synopsis, and that #1844 is the right place to discuss switch-to-last-tab. I will move my discussion to #1844 which already has a near identical implementation for switch-to-last-tab as my comment above.

**Comment by 00-kat** [comment]
One thing of note: there are two major features that are requested here:

  - Most-recently-used tab order switching, or stack-like tab switching.
  - Switching to the last tab, and specifically that.

The former is going to stay here, but **please move the latter to #1844**, thanks! See https://github.com/ghostty-org/ghostty/discussions/2393#discussioncomment-16754516 for rationale.

-----

Discussion about why *implementing MRU switching* may impact those who prefer switching to the last tab can stay here, however. On that note…

Nota bene: switching to the last tab is <strong>not a subset</strong> of MRU switching.

It's actually subtly (but importantly) different: holding down <kbd>ctrl</kbd>+<kbd>tab</kbd> in Firefox, for example, cycles between the tabs in the tab picker UI that shows up:

https://github.com/user-attachments/assets/86d4e8a0-71c7-4cd9-8c0a-a4c8eb68ed1b

Unless there's two or less tabs, since the choice is unambiguous (potential epilepsy warning):

https://github.com/user-attachments/assets/7e856cd4-c2ce-490d-814c-d9abd9dbc37c

However, with switching to the last tab, it's always unambiguous, so there *is* no tab picker UI.

The inability to holding down <kbd>ctrl</kbd>+<kbd>tab</kbd> isn't *too* problematic, but what is problematic is how quickly that tab picker UI shows up. I've used enough software—I believe Windows' <kbd>alt</kbd>+<kbd>tab</kbd> is an example—where if I press <kbd>ctrl</kbd>+<kbd>tab</kbd> but take ever so slightly too long to release <kbd>ctrl</kbd>, the tab UI shows up and then immediately disappears again, which is jarring if you just want “switch to last tab”. For me, Firefox actually does a really good job with its delay: it took me considerable effort to get the jarring effect; but it's very dependent on the user's dexterity, and it's definitely likely that I just have more trouble pressing keys slowly.

**Reply by mnemnion** [reply]
I think, perhaps, the tab picker is not central to the distinction you're drawing. Commenting to try and reflect my own understanding here, and avoid some head-scratching of the "does Ghostty even need a tab picker?" variety.

If `ctrl+tab` means "last tab", then switching between two tabs works like this: `ctrl+tab+tab+tab+tab`.  With MRU, that does something else: but you get "last tab" behavior like so: `ctrl+tab ctrl+tab ctrl+tab`.   The "last tab" version also doesn't really have a use for `ctrl+shift+tab` because it doesn't have directions.

So, not a _subset_ because the natural implementations lead to different command sequences. But, a subset in the sense that MRU can be used in a last-tab-esque fashion, just not in the _same_ fashion.

Is that where you're going with this?

**Reply by 00-kat** [reply]
Yeah that's what I mean. They're used differently, I just mean somewhat-subset in implementation, which you stated better here:

> But, a subset in the sense that MRU can be used in a last-tab-esque fashion, just not in the same fashion.

(I mostly said this whole thing because many proponents of MRU said that it subsumes last-tab, and hence said MRU is the better feature to implement. It was sort of a justification as to why they should be separate features.)

**Reply by 00-kat** [reply]
Oh and:

> I think, perhaps, the tab picker is not central to the distinction you're drawing.

I put most of the focus on the tab switcher because it's the most visual way in which the differing behavior bites, and because the way I use MRU is entirely based on muscle memory and I completely forgot about how I also do <kbd>ctrl</kbd>+<kbd>tab</kbd>+<kbd>tab</kbd>+<kbd>tab</kbd>+<kbd>tab</kbd> to switch between two tabs, oops…

**Comment by 00-kat** [comment]
xref #4904, there's some more discussion there.

**Reply by jkinred** [reply]
To help people find the right place, do you think that #4904 is a duplicate of #1844 rather than this discussion?

**Reply by 00-kat** [reply]
Nope, they deliberate a tab *stack*. so that's here.

**Comment by Zowx** [comment]
Adding a real-world use case to support this:

I run several Claude Code agents in parallel across separate tabs, typically 4 to 6 sessions, one per project or context. The positions of these tabs change throughout the day as I open and close sessions, so `cmd+1..9` is unreliable for the "jump back to where I was" gesture. I end up scanning the tab bar to count positions, which breaks flow constantly.

The two-tab toggle from #1844 would help, but it's not quite enough for this workflow. The cognitive pattern is exactly what @dmik described above: "bring me back to what I was doing before, then bring me back here." With 4 to 6 active contexts, that's a stack, not just a pair.

For what it's worth: I tried switching from iTerm2 to Ghostty this week. Render performance and startup are noticeably better, but the missing MRU `Ctrl+Tab` is what pushed me back within a day. I suspect other power users juggling multiple long-running CLI agents are hitting the same wall.

Happy to test a preview build whenever the design lands, even if the binding details are still up for debate.

**Comment by elektronik2k5** [comment]
Agreed - MRU is very important and its lack is a blocker for adopting Ghostty by power users.

**Comment by AryanilPanja** [comment]
Will this feature be added? This is one of the most important features to me to be more efficient. I don't want to hold shift along with ctrl-tab to go to the "previous" (not even last used) tab and if I am working on two tabs i would like to switch between then by a simple alt+tab. At least it can be added as a toggle which can be enabled by users who prefer it that way.

**Reply by evmcl** [reply]
Agreed. This is really the only thing keeping me on Konsole (even though I don't run full KDE). It doesn't seem any of the newer terminals support this. (Ghostty, Kitty and Wezterm are the ones I checked.) Feels like a small opportunity here…
